src/dmd/inlinecost.d

```diff
@@ -216,42 +219,49 @@ public:
             return;
         }
         expressionInlineCost(s.condition);
-        /* Specifically allow:
-         *  if (condition)
-         *      return exp1;
-         *  else
-         *      return exp2;
-         * Otherwise, we can't handle return statements nested in if's.
-         */
-        if (s.elsebody && s.ifbody && s.ifbody.isReturnStatement() && s.elsebody.isReturnStatement())
-        {
+
+        if (s.ifbody)
             s.ifbody.accept(this);
+        bool thenReturn = foundReturn;
+        foundReturn = false;
+
+        if (s.elsebody)
             s.elsebody.accept(this);
-            //printf("cost = %d\n", cost);
+        bool elseReturn = foundReturn;
+        foundReturn = false;
+
+        if (thenReturn != elseReturn && visitSubseq)
+        {
+            /* Recognize:
+             *  if (condition) { ...; } else { ...; return exp2; }
+             *  ...; return exp1;   // subsequent of 'then'.
+             *
+             * and:
+             *  if (condition) { ...; return exp1; } else { ...; }
+             *  ...; return exp2;   // subsequent of 'else'.
+             */
+            scope InlineCostVisitor v = new InlineCostVisitor();
+            visitSubseq(v);
+            cost += v.cost;
+            if (!thenReturn)
+                thenReturn = v.foundReturn;
```

Comment:
need coverage test

Reply:
Adding a third test where there is no return in the if branch.
```
pragma(inline, true)
{
    int foo3(int v)
    {
        return bar3(2 * v);
    }

    int bar3(int a)
    {
        if (a > 0)
            a = 1;
        else
            return baz3(a);
        return a;
    }
}

int baz3(int a)
{
    if (a > 0)
        throw new Exception("a > 0");
    return a - 1;
}
```
Check latest dmd release, and it can't inline it.
```
inline7625.d(62): Error: function inline7625.bar3 cannot inline function
```

```